cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll

```diff
@@ -658,14 +658,18 @@ newtype TTranslatedElement =
         t instanceof ReferenceType
       ) and
       (
-        isWrite = true or
+        isWrite = true and
+        not call.getTarget().getParameter(n).getType().isDeeplyConstBelow()
+        or
         isWrite = false
       )
       or
       not call.getTarget() instanceof SideEffectFunction and
       n = -1 and
       (
-        isWrite = true or
+        isWrite = true and
+        not call.getTarget() instanceof ConstMemberFunction
```

Comment:
We have `isDeeplyConstBelow` in the parameter case. Do we need something similar here to account for fields containing pointers to non-const?

Reply:
I think that case is covered by ensuring that the callee isn't a const member function since a call to a non-const member function implies that the qualifier isn't const (I think?). Or am I misunderstanding the case you're thinking of?

Reply:
I was thinking something like `setStarP` below, where we might want a `PostUpdateNode` on the qualifier to allow field flow to handle `*p` in a sufficiently-glorious future. After looking at it again, `isDeeplyConstBelow` won't handle that case either, so this would be unnecessary futureproofing.
```
class MyIntPointer {
  int *p;
  
  void setStarP(int x) {
    *p = x;
  }
}
```

Reply:
Oh! Hm, yes I can see that we do actually want to have write side effects for these cases.

Reply:
I've talked to Jonas about this issue and we agreed that this is a good change for the IR in general, and we'll fix this for dataflow specifically when it becomes a problem. It'll probably involve adding a new dataflow node.